File: VB/StylesContainerControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraCharts.Native

Namespace DevExpress.XtraCharts.Design
	Public Class StylesContainerControl
		Inherits XtraScrollableControl
		Private Class AppearanceColors
			Private appearance_Renamed As RootAppearance
			Private palette As Palette
			Private editors() As StyleEdit
			Public ReadOnly Property Appearance() As RootAppearance
				Get
					Return appearance_Renamed
				End Get
			End Property
			Public ReadOnly Property ColorsCount() As Integer
				Get
					Return palette.Count + 1
				End Get
			End Property
			Default Public Property Item(ByVal index As Integer) As StyleEdit
				Get
					Return editors(index)
				End Get
				Set(ByVal value As StyleEdit)
					editors(index) = value
				End Set
			End Property
			Public Sub New(ByVal appearance As RootAppearance, ByVal palette As Palette)
				Me.appearance_Renamed = appearance
				Me.palette = palette
				editors = New StyleEdit(ColorsCount - 1){}
			End Sub
		End Class
		Private paletteRepository As PaletteRepository
		Private matrix As List(Of AppearanceColors)
		Private lastRegisteredAppearance As Integer
		Private elementSize As Size
		Private current As Point
		Private chart As Chart
		Private lockChangeCurrent As Boolean = False
		Public ReadOnly Property CurrentAppearance() As RootAppearance
			Get
				Return matrix(current.Y).Appearance
			End Get
		End Property
		Public ReadOnly Property CurrentPaletteIndex() As Integer
			Get
				Return current.X
			End Get
		End Property
		Public Event OnEditValueChanged As EventHandler
		Public Event OnNeedClose As EventHandler
		Public Sub Initialize(ByVal chart As Chart)
			Me.chart = chart
			UpdateIt()

		End Sub

		Public Sub UpdateIt()
			DisposeAppearanceImages()
			Dim viewType As ViewType
			If chart.Series.Count > 0 Then
				viewType = SeriesViewFactory.GetViewType(chart.Series(0).View)
			Else
				viewType = ViewType.Bar
			End If
			SuspendLayout()
			Try
				SetPaletteRepository(chart.PaletteRepository)
				SetAppearancesCount(chart.AppearanceRepository.Names.Length)
				For Each appearance As RootAppearance In chart.AppearanceRepository
					If (Not chart.Palette.Predefined) OrElse appearance Is chart.Appearance OrElse String.IsNullOrEmpty(appearance.PaletteName) OrElse appearance.PaletteName = chart.Palette.Name Then
						RegisterAppearance(appearance, viewType, chart.Palette)
					End If
				Next appearance
			Finally
				ResumeLayout()
			End Try
		End Sub
		Public Sub SetPaletteRepository(ByVal paletteRepository As PaletteRepository)
			Me.paletteRepository = paletteRepository
		End Sub
		Public Sub SetAppearancesCount(ByVal count As Integer)
			matrix = New List(Of AppearanceColors)(count)
			lastRegisteredAppearance = 0
		End Sub
		Public Sub SelectStyle(ByVal appearance As RootAppearance, ByVal paletteIndex As Integer)
			For i As Integer = 0 To matrix.Count - 1
				Dim row As AppearanceColors = matrix(i)
				If row.Appearance Is appearance AndAlso paletteIndex < row.ColorsCount Then
					current = New Point(paletteIndex, i)
					lockChangeCurrent = True
					Return
				End If
			Next i
			current = Point.Empty
		End Sub
		Public Sub RegisterAppearance(ByVal appearance As RootAppearance, ByVal viewType As ViewType, ByVal palette As Palette)
			Dim row As New AppearanceColors(appearance, palette)
			matrix.Add(row)
			For i As Integer = 0 To palette.Count
				row(i) = AddStyleEditor(appearance.CreateImage(viewType, palette, i), i)
			Next i
			lastRegisteredAppearance += 1
		End Sub
		Private Sub DisposeAppearanceImages()
			For Each control As Control In Controls
				Dim pictureEdit As PictureEdit = TryCast(control, PictureEdit)
				If pictureEdit IsNot Nothing AndAlso pictureEdit.Image IsNot Nothing Then
					pictureEdit.Image.Dispose()
				End If
				control.Dispose()
			Next control
			Controls.Clear()
		End Sub
		Private Function AddStyleEditor(ByVal image As Image, ByVal column As Integer) As StyleEdit
			elementSize = image.Size
			Dim styleEdit As New StyleEdit()
			styleEdit.Location = New Point(elementSize.Width * column, elementSize.Height * lastRegisteredAppearance)
			styleEdit.Size = elementSize
			styleEdit.Image = image
			styleEdit.BorderStyle = BorderStyles.NoBorder
			styleEdit.Properties.ReadOnly = True
			Controls.Add(styleEdit)
			Return styleEdit
		End Function
		Private Sub SelectCurrentChild()
			lockChangeCurrent = True
			Try
				matrix(current.Y)(current.X).Focus()
			Finally
				lockChangeCurrent = False
			End Try
		End Sub
		Private Sub CheckCurrentX()
			Dim colorsCount As Integer = matrix(current.Y).ColorsCount
			If current.X >= colorsCount Then
				current.X = colorsCount - 1
			End If
		End Sub
		Private Sub EditValueChanged()
			Dim palette As Palette = chart.Palette
			chart.Appearance = CurrentAppearance
			chart.Palette = palette
			chart.PaletteBaseColorNumber = CurrentPaletteIndex
			RaiseEvent OnEditValueChanged(Me, EventArgs.Empty)

		End Sub
		Friend Sub RaiseNeedClose()
			RaiseEvent OnNeedClose(Me, EventArgs.Empty)
		End Sub
		Friend Sub OnFocusChanged()
			If lockChangeCurrent Then
				Return
			End If
			For y As Integer = 0 To matrix.Count - 1
				Dim row As AppearanceColors = matrix(y)
				For x As Integer = 0 To row.ColorsCount - 1
					If row(x).Focused Then
						current.X = x
						current.Y = y
						EditValueChanged()
						Return
					End If
				Next x
			Next y
		End Sub
		Protected Overrides Sub OnGotFocus(ByVal e As EventArgs)
			SelectCurrentChild()
		End Sub
		Protected Overrides Function ProcessDialogKey(ByVal keyData As Keys) As Boolean
			Select Case keyData
				Case Keys.Left
'INSTANT VB TODO TASK: Assignments within expressions are not supported in VB.NET
'ORIGINAL LINE: if (--current.X < 0)
					If --current.X < 0 Then
						current.X = matrix(current.Y).ColorsCount - 1
					End If
				Case Keys.Right
'INSTANT VB TODO TASK: Assignments within expressions are not supported in VB.NET
'ORIGINAL LINE: if (++current.X >= matrix[current.Y].ColorsCount)
					If ++current.X >= matrix(current.Y).ColorsCount Then
						current.X = 0
					End If
				Case Keys.Up
'INSTANT VB TODO TASK: Assignments within expressions are not supported in VB.NET
'ORIGINAL LINE: if (--current.Y < 0)
					If --current.Y < 0 Then
						current.Y = matrix.Count - 1
					End If
					CheckCurrentX()
				Case Keys.Down
'INSTANT VB TODO TASK: Assignments within expressions are not supported in VB.NET
'ORIGINAL LINE: if (++current.Y >= matrix.Count)
					If ++current.Y >= matrix.Count Then
						current.Y = 0
					End If
					CheckCurrentX()
				Case Keys.Home
					current.X = 0
					current.Y = 0
				Case Keys.End
					current.Y = matrix.Count - 1
					current.X = matrix(current.Y).ColorsCount - 1
				Case Keys.PageUp
					Dim [step] As Integer = ClientSize.Height \ elementSize.Height
					If [step] = 0 Then
						[step] += 1
					End If
					current.Y -= [step]
					If current.Y < 0 Then
						current.Y = 0
					End If
					current.X = 0
					Exit Select
				Case Keys.PageDown
					Dim [step] As Integer = ClientSize.Height \ elementSize.Height
					If [step] = 0 Then
						[step] += 1
					End If
					current.Y += [step]
					If current.Y >= matrix.Count Then
						current.Y = matrix.Count - 1
					End If
					current.X = Integer.MaxValue
					CheckCurrentX()
					Exit Select
				Case Else
					Return MyBase.ProcessDialogKey(keyData)
			End Select
			SelectCurrentChild()
			EditValueChanged()
			Return True
		End Function
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			MyBase.Dispose(disposing)
			DisposeAppearanceImages()
		End Sub
	End Class
End Namespace

File: CS/StylesContainerControl.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraCharts.Native;

namespace DevExpress.XtraCharts.Design {
    public class StylesContainerControl : XtraScrollableControl {
        class AppearanceColors {
            RootAppearance appearance;
            Palette palette;
            StyleEdit[] editors;
            public RootAppearance Appearance { get { return appearance; } }
            public int ColorsCount { get { return palette.Count + 1; } }
            public StyleEdit this[int index] {
                get { return editors[index]; }
                set { editors[index] = value; }
            }
            public AppearanceColors(RootAppearance appearance, Palette palette) {
                this.appearance = appearance;
                this.palette = palette;
                editors = new StyleEdit[ColorsCount];
            }
        }
        PaletteRepository paletteRepository;
        List<AppearanceColors> matrix;
        int lastRegisteredAppearance;
        Size elementSize;
        Point current;
        Chart chart;
        bool lockChangeCurrent = false;
        public RootAppearance CurrentAppearance { get { return matrix[current.Y].Appearance; } }
        public int CurrentPaletteIndex { get { return current.X; } }
        public event EventHandler OnEditValueChanged;
        public event EventHandler OnNeedClose;
        public void Initialize(Chart chart) {
            this.chart = chart;
            UpdateIt();
        
        }

		public void UpdateIt() {
			DisposeAppearanceImages();
			ViewType viewType = chart.Series.Count > 0 ?
				SeriesViewFactory.GetViewType(chart.Series[0].View) : ViewType.Bar;
			SuspendLayout();
			try {
				SetPaletteRepository(chart.PaletteRepository);
				SetAppearancesCount(chart.AppearanceRepository.Names.Length);
				foreach (RootAppearance appearance in chart.AppearanceRepository)
					if (!chart.Palette.Predefined || appearance == chart.Appearance || String.IsNullOrEmpty(appearance.PaletteName) || appearance.PaletteName == chart.Palette.Name)
						RegisterAppearance(appearance, viewType, chart.Palette);
			}
			finally {
				ResumeLayout();
			}
		}
        public void SetPaletteRepository(PaletteRepository paletteRepository) {
            this.paletteRepository = paletteRepository;
        }
        public void SetAppearancesCount(int count) {
            matrix = new List<AppearanceColors>(count);
            lastRegisteredAppearance = 0;
        }
        public void SelectStyle(RootAppearance appearance, int paletteIndex) {
            for (int i = 0; i < matrix.Count; i++) {
                AppearanceColors row = matrix[i];
                if (row.Appearance == appearance && paletteIndex < row.ColorsCount) {
                    current = new Point(paletteIndex, i);
                    lockChangeCurrent = true;
                    return;
                }
            }
            current = Point.Empty;
        }
        public void RegisterAppearance(RootAppearance appearance, ViewType viewType, Palette palette) {
            AppearanceColors row = new AppearanceColors(appearance, palette);
            matrix.Add(row);
            for (int i = 0; i <= palette.Count; i++)
                row[i] = AddStyleEditor(appearance.CreateImage(viewType, palette, i), i);
            lastRegisteredAppearance++;
        }
		void DisposeAppearanceImages() {
			foreach (Control control in Controls) {
				PictureEdit pictureEdit = control as PictureEdit;
				if (pictureEdit != null && pictureEdit.Image != null)
					pictureEdit.Image.Dispose();
				control.Dispose();
			}
			Controls.Clear();
		}
        StyleEdit AddStyleEditor(Image image, int column) {
            elementSize = image.Size;
            StyleEdit styleEdit = new StyleEdit();
            styleEdit.Location = new Point(elementSize.Width * column, elementSize.Height * lastRegisteredAppearance);
            styleEdit.Size = elementSize;
            styleEdit.Image = image;
            styleEdit.BorderStyle = BorderStyles.NoBorder;
            styleEdit.Properties.ReadOnly = true;
            Controls.Add(styleEdit);
            return styleEdit;
        }
        void SelectCurrentChild() {
            lockChangeCurrent = true;
            try {
                matrix[current.Y][current.X].Focus();
            }
            finally {
                lockChangeCurrent = false;
            }
        }
        void CheckCurrentX() {
            int colorsCount = matrix[current.Y].ColorsCount;
            if (current.X >= colorsCount)
                current.X = colorsCount - 1;
        }
        void EditValueChanged() {
            Palette palette = chart.Palette;
            chart.Appearance = CurrentAppearance;
            chart.Palette = palette;
            chart.PaletteBaseColorNumber = CurrentPaletteIndex;
            if (OnEditValueChanged != null)
                OnEditValueChanged(this, EventArgs.Empty);
            
        }
        internal void RaiseNeedClose() {
            if (OnNeedClose != null)
                OnNeedClose(this, EventArgs.Empty);
        }
        internal void OnFocusChanged() {
            if (lockChangeCurrent)
                return;
            for (int y = 0; y < matrix.Count; y++) {
                AppearanceColors row = matrix[y];
                for (int x = 0; x < row.ColorsCount; x++) 
                    if (row[x].Focused) {
                        current.X = x;
                        current.Y = y;
                        EditValueChanged();
                        return;
                    }
            }
        }
        protected override void OnGotFocus(EventArgs e) {
            SelectCurrentChild();
        }
        protected override bool ProcessDialogKey(Keys keyData) {
            switch (keyData) {
                case Keys.Left:
                    if (--current.X < 0)
                        current.X = matrix[current.Y].ColorsCount - 1;
                    break;
                case Keys.Right:
                    if (++current.X >= matrix[current.Y].ColorsCount)
                        current.X = 0;
                    break;
                case Keys.Up:
                    if (--current.Y < 0)
                        current.Y = matrix.Count - 1;
                    CheckCurrentX();
                    break;
                case Keys.Down:
                    if (++current.Y >= matrix.Count)
                        current.Y = 0;
                    CheckCurrentX();
                    break;
                case Keys.Home:
                    current.X = 0;
                    current.Y = 0;
                    break;
                case Keys.End:
                    current.Y = matrix.Count - 1;
                    current.X = matrix[current.Y].ColorsCount - 1;
                    break;
                case Keys.PageUp: {
                    int step = ClientSize.Height / elementSize.Height;
                    if (step == 0)
                        step++;
                    current.Y -= step;
                    if (current.Y < 0)
                        current.Y = 0;
                    current.X = 0;
                    break;
                }
                case Keys.PageDown: {
                    int step = ClientSize.Height / elementSize.Height;
                    if (step == 0)
                        step++;
                    current.Y += step;
                    if (current.Y >= matrix.Count)
                        current.Y = matrix.Count - 1;
                    current.X = int.MaxValue;
                    CheckCurrentX();
                    break;
                }
                default:
                    return base.ProcessDialogKey(keyData);
            }
            SelectCurrentChild();
            EditValueChanged();
            return true;
        }
		protected override void Dispose(bool disposing) {
			base.Dispose(disposing);
			DisposeAppearanceImages();
		}
    }
}
